components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -718,13 +718,56 @@ private boolean isDomainNameEmbeddedInAttributeValue(String filterOperation, Str
             String[] userStoreDomainNames = getDomainNames();
             ArrayList<String> filteredUsernames = new ArrayList<>();
             for (String userStoreDomainName : userStoreDomainNames) {
-                filteredUsernames.addAll(Arrays
-                        .asList(getFilteredUsernames(node, offset, limit, sortBy, sortOrder, userStoreDomainName)));
+                String[] userNames = getFilteredUsernames(node, offset, limit, sortBy, sortOrder, userStoreDomainName);
+
+                // Calculating new offset and limit parameters.
+                int[] newParams = calculateOffsetAndLimit(node, offset, limit, sortBy, sortOrder, domainName,
+                        userNames.length);
+                offset = newParams[0];
+                limit = newParams[1];
+                filteredUsernames.addAll(Arrays.asList(userNames));
+
+                // If the limit is changed then filtering needs to be stopped.
+                if (limit <= 0)
+                    break;
             }
             return filteredUsernames.toArray(new String[0]);
         }
     }
 
+    /**
+     * Method to update the offset and count(limit) when iterating a filter across all domains.
+     *
+     * @param node                  Expression node
+     * @param offset                Starting index
+     * @param limit                 Countinig value
+     * @param sortBy                Sort by
+     * @param sortOrder             Sort order
+     * @param domainName            Domain to be filtered
+     * @param numberOfFilteredUsers Amount of users filtered in the search
+     * @return Array with newly updated offset and limit
+     * @throws CharonException Error while filtering the domain from index 1 to offset
+     */
+    private int[] calculateOffsetAndLimit(Node node, int offset, int limit, String sortBy, String sortOrder,
+            String domainName, int numberOfFilteredUsers) throws CharonException {
+
+        // Filtered user names of size zero implies either it has any matches for the filter or number of matched
+        // users is less than the offset.
+        if (numberOfFilteredUsers <= 0 && offset > 1) {
```

Comment:
shall we add debug logs covering both cases

Reply:
fixed: 034f63e984cc91211fb945bc398451d985fa7f51